### PR TITLE
Linting and starting to use Tox

### PR DESCRIPTION
**Pull request by agjohnson**
This is a example of some improved linting and testing configuration, thoughts are welcome.

This is an example of linting using [Prospector][prospector], which is a wrapper around a number of tools -- including flake8 and pylint. It includes a configuration that will set a low bar to getting back to linting code, and is configurable by strictness profile -- which is something we can increase when we start to clean things up more.

Output of a run here:
https://travis-ci.org/rtfd/readthedocs.org/builds/71619840

This also starts to use tox, as suggested by @destroyerofbuilds, which seems like an easy way to at least combine our testing, linting, and doc creation test commands into one tool, instead of one-off shell scripts. It makes for more reproducible tests as well, handling virtualenv creation.

This also fixes a bug on dict ordering in a test case.

[prospector]: https://github.com/landscapeio/prospector

### CODE REVIEW

File: tox.ini [hunk 1]
```diff
@@ -0,0 +1,28 @@
+[tox]
+envlist = py26,py27,lint,docs
+skipsdist = True
+
+[testenv]
+setenv =
+    PYTHONPATH={toxinidir}/readthedocs:{toxinidir}
+    DJANGO_SETTINGS_MODULE=settings.test
+deps = -r{toxinidir}/requirements/pip.txt
```
Comment (hbetts) [inline tox.ini]: I think, based on [depending on requitements.txt](https://testrun.org/tox/latest/example/basic.html#depending-on-requirements-txt) that all commands are already executed with `{toxinidir}` as the working directory, making the inclusion of it in your paths redundant.

File: .travis.yml [hunk 2]
```diff
@@ -1,13 +1,15 @@
 language: python
 python:
- - "2.7"
+ - 2.7
+env:
```
Comment (hbetts) [inline .travis.yml]: Would it be alright to add blank lines before each life cycle stage? I think it would make the file contents a little more readable, and conform more closely with the style of the other configuration files.

File: tox.ini [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+    py.test {posargs}
+
+[testenv:docs]
+changedir = {toxinidir}/docs
+commands =
+    sphinx-build -b html -d {envtmpdir}/doctrees .  {envtmpdir}/html
+
+[testenv:lint]
+deps =
+    {[testenv]deps}
```
Comment (hbetts) [inline tox.ini]: Do you really need your entire project's set of dependencies for linting?

Comment (agjohnson) [inline tox.ini]: Yes, otherwise there will be dependency failures from pylint.

File: .travis.yml (hunk 2 above)
Comment (agjohnson) [inline .travis.yml]: I feel like that is extraneous whitespace.

File: tox.ini (hunk 1 above)
Comment (agjohnson) [inline tox.ini]: Explicit is better than implicit here.

File: tox.ini [hunk 4]
```diff
@@ -0,0 +1,28 @@
+[tox]
+envlist = py26,py27,lint,docs
```
Comment (gregmuellegger) [inline tox.ini]: Why are we testing against py26 here? Do we really support it? Django 1.8 requires at least Python 2.7 https://docs.djangoproject.com/en/dev/releases/1.8/

File: tox.ini [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+skipsdist = True
+
+[testenv]
+setenv =
+    PYTHONPATH={toxinidir}/readthedocs:{toxinidir}
+    DJANGO_SETTINGS_MODULE=settings.test
+deps = -r{toxinidir}/requirements/pip.txt
+changedir = {toxinidir}/readthedocs
+commands =
+    py.test {posargs}
```
Comment (gregmuellegger) [inline tox.ini]: in `runtests.sh` we use `LANG=C coverage run -m pytest`. The reason for the LANG=C is that if you don't have a english locale set in the system, the tests will fail with this assertion:

```
E       AssertionError: Regexp didn't match: '(?:No such file or directory|not found)' not found in '/bin/sh: /non-existant/foobar: Datei oder Verzeichnis nicht gefunden\n'
```

We should use `LANG=C` with tox as well.

File: tox.ini (hunk 4 above)
Comment (agjohnson) [inline tox.ini]: I dropped py26 in there as we've so far maintained backwards compatibility, though we aren't running tests on travis for py26. I didn't realize 1.7 dropped py26 support though, so this can likely be dropped

File: tox.ini [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+
+[testenv:docs]
+changedir = {toxinidir}/docs
+commands =
+    sphinx-build -b html -d {envtmpdir}/doctrees .  {envtmpdir}/html
```
Comment (gregmuellegger) [inline tox.ini]: A big :+1: on testing the doc build. I recommend using the `-W` which will turn warnings into errors to make sure we don't have unhandled warnings in the docs.

File: readthedocs/rtd_tests/tests/test_project_symlinks.py [hunk 7]
```diff
@@ -58,10 +58,12 @@ def test_symlink_basic(self):
...
-        for (i, command) in enumerate(commands):
-            self.assertEqual(self.commands[i], command.format(**self.args),
-                             msg=('Command {0} mismatch, expecting {1}'
-                                  .format(i, self.commands[i])))
+
+        for command in commands:
+            self.assertIsNotNone(
+                self.commands.pop(
+                    self.commands.index(command.format(**self.args))
+                ))
```
Comment (gregmuellegger) [inline readthedocs/rtd_tests/tests/test_project_symlinks.py]: How are these changes related to the introduction of tox?

Comment (agjohnson) [inline readthedocs/rtd_tests/tests/test_project_symlinks.py]: Tox introduces `PYTHONHASHSEED` env variable to testing, which caught these bugs as failing due to dictionary ordering inconsistencies. This is included here to keep the testing experience the same as before the PR.

File: tox.ini (hunk 6 above)
Comment (agjohnson) [inline tox.ini]: Yes, indeed. I dropped the option as we have some warnings now, I figured this could be addressed with another PR later

File: .travis.yml [hunk 8]
```diff
@@ -1,13 +1,15 @@
 language: python
 python:
- - "2.7"
+ - 2.7
```
Comment (destroyerofbuilds) [inline .travis.yml]: What does listing the Python version here do? I've seen some projects list a Python version, while others leave it out entirely, instead relying on the `TOX_ENV` variable to construct the build matrix.

File: prospector.yml [hunk 9]
```diff
@@ -0,0 +1,30 @@
...
+        max-line-length: 100
+
+pylint:
+  max-line-length: 100
+
+mccabe:
+  run: false
+
+pep257:
+  run: false
```
Comment (destroyerofbuilds) [inline prospector.yml]: What are your thoughts on enabling the optional linter `vulture`?

**Comment by ericholscher** [comment]
Seems odd that the tests are passing on this PR. We definitely have lines >100 in length?

**Comment by ericholscher** [comment]
:+1: on this -- we should fix all the errors before merging though, perhaps in another PR on top?

**Comment by agjohnson** [comment]
I'll rebase this up and add some docs. I'll leave the `runtests.sh` for now, but the docs will reflect running tox locally. I'll open up some tickets against this as well, so some of these issues can be followed up on.

**Comment by agjohnson** [comment]
Docs added, this should be ready to go if it looks good.

@ericholscher I added some follow up tickets to address linting and enabling them with travis. Safe to run test suite this way for now and follow up work on other PRs without blocking getting this merged.

**Comment by ericholscher** [comment]
Ship it!
